File: Tree/114.Flatten_Binary_Tree_to_Linked_List/114.Flatten_Binary_Tree_to_Linked_List.cpp
// Recursive
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *cur;
    void flatten(TreeNode* root) {
        if (!root) return;

        cur = root;
        flatten(root->left);

        TreeNode *temp = root->right;
        root->right = root->left;
        root->left = NULL;
        cur->right = temp;
        
        flatten(temp);
    }
};

File: Tree/114.Flatten_Binary_Tree_to_Linked_List/114.Flatten_Binary_Tree_to_Linked_List.py
# Iterate

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root :
            return
        
        p = root
        while p :
            if p.left :
                temp = p.right
                p.right = p.left
                p.left = None
                
                maxleft = p
                while maxleft.right :
                    maxleft = maxleft.right
                maxleft.right = temp
                
            p = p.right